File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/verbs/validate.py
from __future__ import annotations

from pathlib import Path

from ..cli_format import format_human, format_json
from ..parser import (
    MalformedFrontmatterError,
    MissingFrontmatterError,
    parse_frontmatter,
)
from ..rules.body import check_body
from ..rules.naming import check_naming
from ..rules.optional_dirs import check_optional_dirs
from ..rules.references import check_references
from ..types import Issue, Report


def _discover(target: Path) -> list[Path]:
    if target.is_file():
        if target.name == "SKILL.md":
            return [target.parent]
        return []
    direct = target / "SKILL.md"
    if direct.is_file():
        return [target]
    out: list[Path] = []
    for entry in sorted(target.iterdir()):
        if not entry.is_dir() or entry.name.startswith("."):
            continue
        out.extend(_discover(entry))
    return out


def _validate_one(skill_dir: Path, extra_tiers: list[str]) -> Report:
    issues: list[Issue] = []
    skill_path = skill_dir / "SKILL.md"
    try:
        text = skill_path.read_text(encoding="utf-8")
    except OSError as e:
        issues.append(
            Issue("E100_SKILL_MD_MISSING", "error", "skill", f"SKILL.md not readable: {e}")
        )
        return _build(skill_path, issues)
    try:
        parsed = parse_frontmatter(text)
    except MissingFrontmatterError as e:
        issues.append(Issue("E101_FRONTMATTER_MISSING", "error", "frontmatter", str(e)))
        return _build(skill_path, issues)
    except MalformedFrontmatterError as e:
        issues.append(
            Issue("E102_FRONTMATTER_MALFORMED", "error", "frontmatter", str(e))
        )
        return _build(skill_path, issues)
    issues.extend(check_naming(parsed, skill_dir.name, {"extra_tiers": extra_tiers}))
    issues.extend(check_body(parsed))
    issues.extend(check_references(parsed, str(skill_dir)))
    issues.extend(check_optional_dirs(str(skill_dir)))
    return _build(skill_path, issues)


def _build(path: Path, issues: list[Issue]) -> Report:
    errors = sum(1 for i in issues if i.severity == "error")
    warnings = sum(1 for i in issues if i.severity == "warn")
    status = "FAIL" if errors > 0 else ("WARN" if warnings > 0 else "PASS")
    return Report(
        path=str(path),
        status=status,
        issues=issues,
        summary={"errors": errors, "warnings": warnings},
    )


def validate(args) -> tuple[int, str]:
    target = Path(args.path).resolve()
    extra = [t.strip() for t in (args.extra_tiers or "").split(",") if t.strip()]
    skills = _discover(target)
    if not skills:
        return 1, f"no SKILL.md found under: {args.path}\n"
    reports = [_validate_one(s, extra) for s in skills]
    total_errors = sum(r.summary["errors"] for r in reports)
    total_warnings = sum(r.summary["warnings"] for r in reports)
    if total_errors > 0:
        exit_code = 1
    elif args.strict and total_warnings > 0:
        exit_code = 2
    else:
        exit_code = 0
    if args.format == "json":
        if len(reports) == 1:
            out = format_json(reports[0])
        else:
            top = {
                "status": (
                    "FAIL" if total_errors > 0
                    else ("WARN" if total_warnings > 0 else "PASS")
                ),
                "reports": [r.to_dict() for r in reports],
                "summary": {"errors": total_errors, "warnings": total_warnings},
            }
            out = format_json(top)
    else:
        out = "".join(format_human(r, color=not args.no_color) for r in reports)
    return exit_code, out

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/cli/verbs/validate.mjs
import { readFileSync, statSync, readdirSync } from "node:fs";
import { join, basename, resolve } from "node:path";
import {
  parseFrontmatter,
  MissingFrontmatterError,
  MalformedFrontmatterError,
} from "../../parser.mjs";
import { checkNaming } from "../../rules/naming.mjs";
import { checkBody } from "../../rules/body.mjs";
import { checkReferences } from "../../rules/references.mjs";
import { checkOptionalDirs } from "../../rules/optional_dirs.mjs";
import { makeIssue } from "../../types.mjs";
import { formatHuman, formatJson } from "../format.mjs";

function discoverSkills(target) {
  const st = statSync(target);
  if (st.isFile()) {
    if (basename(target) === "SKILL.md") return [resolve(target, "..")];
    return [];
  }
  if (basename(target).toLowerCase() === "skill.md") return [resolve(target, "..")];
  // Directory: either it directly contains SKILL.md, or recurse one level.
  const direct = join(target, "SKILL.md");
  try {
    if (statSync(direct).isFile()) return [target];
  } catch {
    /* fall through */
  }
  const out = [];
  for (const entry of readdirSync(target, { withFileTypes: true })) {
    if (!entry.isDirectory()) continue;
    if (entry.name.startsWith(".")) continue;
    out.push(...discoverSkills(join(target, entry.name)));
  }
  return out;
}

function validateOne(skillDir, opts) {
  const issues = [];
  const skillPath = join(skillDir, "SKILL.md");
  let text;
  try {
    text = readFileSync(skillPath, "utf8");
  } catch (e) {
    issues.push(
      makeIssue(
        "E100_SKILL_MD_MISSING",
        "error",
        "skill",
        `SKILL.md not readable: ${e.message}`,
      ),
    );
    return buildReport(skillPath, issues);
  }
  let parsed;
  try {
    parsed = parseFrontmatter(text);
  } catch (e) {
    if (e instanceof MissingFrontmatterError) {
      issues.push(
        makeIssue("E101_FRONTMATTER_MISSING", "error", "frontmatter", e.message),
      );
    } else if (e instanceof MalformedFrontmatterError) {
      issues.push(
        makeIssue(
          "E102_FRONTMATTER_MALFORMED",
          "error",
          "frontmatter",
          e.message,
        ),
      );
    } else {
      throw e;
    }
    return buildReport(skillPath, issues);
  }
  const dirName = basename(skillDir);
  issues.push(
    ...checkNaming(parsed, dirName, { extraTiers: opts.extraTiers }),
  );
  issues.push(...checkBody(parsed));
  issues.push(...checkReferences(parsed, skillDir));
  issues.push(...checkOptionalDirs(skillDir));
  return buildReport(skillPath, issues);
}

function buildReport(path, issues) {
  let errors = 0;
  let warnings = 0;
  for (const i of issues) {
    if (i.severity === "error") errors++;
    else warnings++;
  }
  const status = errors > 0 ? "FAIL" : warnings > 0 ? "WARN" : "PASS";
  return { path, status, issues, summary: { errors, warnings } };
}

function parseArgs(args) {
  const opts = {
    path: null,
    format: "human",
    noColor: false,
    strict: false,
    extraTiers: [],
  };
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === "--format") opts.format = args[++i];
    else if (a === "--no-color") opts.noColor = true;
    else if (a === "--strict") opts.strict = true;
    else if (a === "--extra-tiers")
      opts.extraTiers = (args[++i] ?? "")
        .split(",")
        .map((s) => s.trim())
        .filter(Boolean);
    else if (!a.startsWith("--") && opts.path == null) opts.path = a;
  }
  return opts;
}

export async function validate(args) {
  const opts = parseArgs(args);
  if (!opts.path) {
    return {
      exitCode: 64,
      output: "usage: skills-ref validate <path> [--format json|human]\n",
    };
  }
  const skills = discoverSkills(resolve(opts.path));
  if (skills.length === 0) {
    return {
      exitCode: 1,
      output: `no SKILL.md found under: ${opts.path}\n`,
    };
  }
  const reports = skills.map((d) => validateOne(d, opts));
  let exitCode = 0;
  let totalErrors = 0;
  let totalWarnings = 0;
  for (const r of reports) {
    totalErrors += r.summary.errors;
    totalWarnings += r.summary.warnings;
  }
  if (totalErrors > 0) exitCode = 1;
  else if (opts.strict && totalWarnings > 0) exitCode = 2;
  let output;
  if (opts.format === "json") {
    if (reports.length === 1) output = formatJson(reports[0]);
    else
      output = formatJson({
        status:
          totalErrors > 0 ? "FAIL" : totalWarnings > 0 ? "WARN" : "PASS",
        reports,
        summary: { errors: totalErrors, warnings: totalWarnings },
      });
  } else {
    output = reports
      .map((r) => formatHuman(r, { color: !opts.noColor }))
      .join("");
  }
  return { exitCode, output };
}
